ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/Jumping.hs

```diff
@@ -0,0 +1,661 @@
+            let fragment = jTheirFragment jumpInfo
+            modifyTVar (cschState (handle context)) $ \csState ->
+              csState {csCandidate = fragment, csLatestSlot = Just (AF.headSlot fragment) }
+            writeTVar (cschJumpInfo (handle context)) $ Just jumpInfo
+            case jumperState of
+              LookingForIntersection badJumpInfo ->
+                -- @AF.headPoint fragment@ is in @badFragment@, as the jumper
+                -- looking for an intersection is the only client asking for its
+                -- jumps.
+                lookForIntersection nextJumpVar (AF.headPoint fragment) badJumpInfo
+              Happy ->
+                writeTVar (cschJumping (handle context)) $
+                  Jumper nextJumpVar (AF.headPoint fragment) Happy
+              FoundIntersection{} ->
+                -- Only happy jumpers are asked to jump by the dynamo, and only
+                -- jumpers looking for an intersection are asked to jump by
+                -- themselves.
+                error "processJumpResult: Jumpers in state FoundIntersection shouldn't be further jumping."
+
+          RejectedJump badJumpInfo -> do
+            -- @goodPoint@ is in @jTheirFragment jumpInfo@. If the jump was
+            -- requested by the dynamo, this holds because the dynamo is not
+            -- allowed to rollback before the jumps that it requests.
+            --
+            -- If the jump was requested by the jumper, this holds because the
+            -- jumper is looking for an intersection, and such jumper only asks
+            -- for jumps that meet this condition.
+            lookForIntersection nextJumpVar goodPoint badJumpInfo
+  where
+    -- Avoid redundant constraint "HasHeader blk" reported by some ghc's
+    _ = getHeaderFields @blk
+    -- | Given a good point (where we know we agree with the dynamo) and a bad
+    -- fragment (where we know the tip disagrees with the dynamo), either decide
+    -- that we know the intersection for sure (if the bad point is the successor
+    -- of the good point) or program a jump somewhere in the middle to refine
+    -- those points.
+    --
+    -- PRECONDITION: The good point is in the bad fragment.
+    lookForIntersection nextJumpVar goodPoint badJumpInfo = do
+      let badFragment = jTheirFragment badJumpInfo
+      case snd <$> AF.splitAfterPoint badFragment goodPoint of
+        Nothing ->
+          error "processJumpResult: goodPoint not in badFragment"
+        Just nextFragment -> do
+          let len = AF.length nextFragment
+          if len <= 1 then do
+            -- If the fragment only contains the bad tip, we know the
+            -- intersection is the good point.
+            writeTVar (cschJumping (handle context)) $
+              Jumper nextJumpVar goodPoint $
+                FoundIntersection $ AF.headPoint badFragment
+            findObjector (stripContext context) >>=
+              electNewObjector (stripContext context)
```

Comment:
Now that I think about it some more, this sounds a bit overkill. When we figure out the intersection of a peer, there is really only two cases:

- If there is no objector, the peer becomes the objector.

- If there is an objector, then the objector's intersection with the dynamo is the oldest (possibly equal) of all the intersection of `FoundIntersection` jumpers with the dynamo, so we only need to compare this new `FoundIntersection` jumper's intersection with that of the objector and switch objector if it is strictly older.

This is based on the following invariant that all endpoints respect:

- If there exist `FoundIntersection` jumpers, then there is an objector and the intersection of the objector with the dynamo is older or equal to all the intersections of the `FoundIntersection` jumpers with the dynamo.

Reply:
Tried another fix at b6f08ca.